ports/stm32f4/common-hal/pulseio/PWMOut.c

```diff
@@ -0,0 +1,403 @@
+    if (frequency == 0 || frequency > 6000000) {
+        mp_raise_ValueError(translate("Invalid frequency supplied"));
+        //return PWMOUT_INVALID_FREQUENCY;
+    }
+
+    TIM_TypeDef * TIMx;
+    //TODO: style/safety question, use this instead?
+    //uint8_t tim_num_pins = TIM_PIN_ARRAY_LEN;
+    uint8_t tim_num = sizeof(mcu_tim_pin_list)/sizeof(*mcu_tim_pin_list);
+    //TODO: use an enum to make this prettier?
+    bool tim_chan_taken = false;
+    bool tim_taken_f_mismatch = false;
+    bool var_freq_mismatch = false;
+    bool first_time_setup = true;
+
+    for(uint i = 0; i < tim_num; i++) {
+        //if pin is same
+        if(mcu_tim_pin_list[i].pin == pin) {
+            //check if the timer has a channel active
+            if (reserved_tim[mcu_tim_pin_list[i].tim_index-1] != 0) {
+                //is it the same channel? (or all channels reserved by a var-freq)
+                if(reserved_tim[mcu_tim_pin_list[i].tim_index-1] & 1<<(mcu_tim_pin_list[i].channel_index-1)) {
+                    tim_chan_taken = true;
+                    continue; //keep looking, might be another viable option
+                }
+                //If the frequencies are the same it's ok
+                if(tim_frequencies[mcu_tim_pin_list[i].tim_index-1] != frequency) {
+                    tim_taken_f_mismatch = true;
+                    continue; //keep looking
+                }
+                //you can't put a variable frequency on a partially reserved timer
+                if(variable_frequency) {
+                    var_freq_mismatch = true;
+                    continue; //keep looking
+                }
+                first_time_setup = false; //skip setting up the timer
+            }
+            //No problems taken, so set it up
+            self->tim = &mcu_tim_pin_list[i];
+            break;
+        }
+    }
+
+    //handle valid/invalid timer instance
+    //TODO: why doesn't PWM handle its own value errors like every other module?
+    //The samd value errors don't really fit for STM32.
+    if(self->tim!=NULL) {
+        //create instance
+        TIMx = mcu_tim_banks[self->tim->tim_index-1];
+        //reserve timer/channel
+        if (variable_frequency) {
+            reserved_tim[self->tim->tim_index-1] = 0x0F;
+        } else {
+            reserved_tim[self->tim->tim_index-1] |= 1<<(self->tim->channel_index-1);
+        }
+        tim_frequencies[self->tim->tim_index-1] = frequency;
+    } else { //no match found
+        if (tim_chan_taken) {
+            mp_raise_ValueError(translate("Timer hardware is reserved"));
+            //return PWMOUT_ALL_TIMERS_ON_PIN_IN_USE;
+        } else if (tim_taken_f_mismatch) {
+            mp_raise_ValueError(translate("Frequency mismatch with existing reserved channel"));
+            //return PWMOUT_INVALID_FREQUENCY;
+        } else if (var_freq_mismatch) {
+            mp_raise_ValueError(translate("Cannot vary frequency of a partially reserved timer"));
```

Comment:
This is "No timers left for variable frequency."

Reply:
I don't understand how that message would help the user. If they receive this bug, they need to change their pin to a channel on a separate timer, or halt the existing PWM using that timer. Saying that no timers or channels are left for Var Freq misidentifies the problem.

Reply:
I'd suggest "Cannot vary frequency on an in-use timer"?

Reply:
I see what you are saying. That seems OK. In the other error messages I would say "timer channel", because just "channel" may not mean a lot to the user.